jupyterlab_caip_optimizer/src/components/graphs/parallel_coordinates.tsx

```diff
@@ -0,0 +1,395 @@
+          const withinSelectedMetricRange =
+            lineDataList[selectedMetricForColor] >=
+              axisPropsList[selectedMetricForColor]['sliderMin'] &&
+            lineDataList[selectedMetricForColor] <=
+              axisPropsList[selectedMetricForColor]['sliderMax'];
+          // normalized for the slider range
+          return withinSelectedMetricRange
+            ? d3.interpolateRdBu(
+                ((lineDataList[selectedMetricForColor] -
+                  axisPropsList[selectedMetricForColor]['sliderMin']) /
+                  (axisPropsList[selectedMetricForColor]['sliderMax'] -
+                    axisPropsList[selectedMetricForColor]['sliderMin'])) *
+                  1.0
+              )
+            : 'rgba(0, 0, 0, 0.1)'; // light gray colored if not within the selected range of metric axis
+        }
+        return 'rgba(63, 81, 181, 0.8)';
+      };
+
+      const discontinousOutOfRange = lineDataList => {
+        const axes = Object.keys(axisPropsList);
+        for (const axis of axes) {
+          if (
+            axisPropsList[axis]['type'] === 'CATEGORICAL' ||
+            axisPropsList[axis]['type'] === 'DISCRETE'
+          ) {
+            const thisDataIndex = axisPropsList[axis]['values'].indexOf(
+              lineDataList[axis]
+            );
+            if (
+              thisDataIndex < axisPropsList[axis]['sliderMinIndex'] ||
+              thisDataIndex > axisPropsList[axis]['sliderMaxIndex']
+            ) {
+              return true;
+            }
+          }
+        }
+        return false;
+      };
+
+      const checkLineSelected = lineDataList => {
+        /**
+         * Discontinous parameters (categorical, discrete) have ordinal scales in D3.
+         * Thus, when certain values are out of the axis range, it doesn't get interpolated nicely
+         * and the lines will go through the top of the axis, which may confuse viewwers.
+         * Therefore, we make the lines with out of range data points for discontinous axes to be invisible.
+         */
+        if (discontinousOutOfRange(lineDataList)) return 'rgba (0, 0, 0, 0)';
+        if (selectBoxDetails.initialValue !== 0 && clickedAxisName !== 'none') {
+          const axisType = axisPropsList[clickedAxisName]['type'];
+          let withinRange = false;
+          switch (axisType) {
+            case 'INTEGER':
+            case 'DOUBLE':
+              if (
+                lineDataList[clickedAxisName] >
+                  Math.min(
+                    selectBoxDetails.initialValue,
+                    selectBoxDetails.currentValue
+                  ) &&
+                lineDataList[clickedAxisName] <
+                  Math.max(
+                    selectBoxDetails.initialValue,
+                    selectBoxDetails.currentValue
+                  )
+              )
+                withinRange = true;
+              break;
+            case 'DISCRETE':
+            case 'CATEGORICAL': {
```

Comment:
nit: The { } don't seem necessary does it?

Reply:
We get linting errors from eslint about not including { } when we define variables in the case statement. https://eslint.org/docs/rules/no-case-declarations